### DISCUSSION TITLE
x is y -> object.ReferenceEquals(x, y)

### DISCUSSION THREAD

**Discussion by sretz** [dotnet/csharplang#6241] [Language Ideas]
In c# we can test if x is null using "x is null" which the compiler replaces with object.ReferenceEquals(x, null). We can also do "x is object" to make sure x is not null which the compiler replaces with !object.ReferenceEquals(x, null). I propose adding "x is y" which the compiler would replace with object.ReferenceEquals(x, y) where x and y are both reference types and related to each other.

**Comment by HaloFour** [comment]
The right side of that expression is a pattern and patterns must be constants.  I couldn't find a discussion specifically asking for that to be relaxed, but I did find [this comment](https://github.com/dotnet/csharplang/issues/2850#issuecomment-623179988), as well as this [Reddit response](https://www.reddit.com/r/csharp/comments/jtfsbe/only_constant_value_is_allowed_in_pattern_matching/) from a language design team member:

> I do think it's possible we'll relax this eventually, but we'll need to find a good experience around exhaustiveness, particularly in switch expressions. If you have a non-constant pattern, we can't determine if you're handling all cases, and the experience around that is concerning. We're also slightly concerned by users matching against values that trigger side effects, which would have undefined behavior.
